crates/re_query2/src/latest_at/results.rs

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    #[inline]
+    pub fn get<C: Component>(&self) -> Option<&LatestAtComponentResults> {
+        self.components.get(&C::name())
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    ///
+    /// Returns an error if the component is not present.
+    #[inline]
+    pub fn get_required<C: Component>(&self) -> crate::Result<&LatestAtComponentResults> {
+        if let Some(component) = self.components.get(&C::name()) {
+            Ok(component)
+        } else {
+            Err(DeserializationError::MissingComponent {
+                component: C::name(),
+                backtrace: ::backtrace::Backtrace::new_unresolved(),
+            }
+            .into())
+        }
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    ///
+    /// Returns empty results if the component is not present.
+    #[inline]
+    pub fn get_optional<C: Component>(&self) -> &LatestAtComponentResults {
+        if let Some(component) = self.components.get(&C::name()) {
+            component
+        } else {
+            static DEFAULT: LatestAtComponentResults = LatestAtComponentResults::empty();
+            &DEFAULT
+        }
+    }
+}
+
+impl LatestAtResults {
+    #[doc(hidden)]
+    #[inline]
+    pub fn add(&mut self, component_name: ComponentName, index: (TimeInt, RowId), cell: DataCell) {
+        // NOTE: Since this is a compound API that actually emits multiple queries, the index of the
+        // final result is the most recent index among all of its components, as defined by time
+        // and row-id order.
+        //
+        // TODO(#5303): We have to ignore the cluster key in this piece of logic for backwards compatibility
+        // reasons with the legacy instance-key model. This will go away next.
+        use re_types_core::Loggable as _;
+        if component_name != re_types_core::components::InstanceKey::name()
+            && index > self.compound_index
+        {
+            self.compound_index = index;
+        }
+
+        self.components.insert(
+            component_name,
+            LatestAtComponentResults {
+                index,
+                promise: Some(Promise::new(cell)),
+            },
+        );
+    }
+}
+
+// ---
+
+/// Uncached results for a particular component when using a latest-at query.
+#[derive(Debug, Clone)]
+pub struct LatestAtComponentResults {
+    index: (TimeInt, RowId),
+
+    // Option so we can have a constant default value for `Self` for the optional+empty case.
+    promise: Option<Promise>,
+}
+
+impl Default for LatestAtComponentResults {
+    #[inline]
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl LatestAtComponentResults {
+    #[inline]
+    const fn empty() -> Self {
+        Self {
+            index: (TimeInt::STATIC, RowId::ZERO),
+            promise: None,
+        }
+    }
```

Comment:
slightly a matter of taste, but I think this reads nicer

```suggestion
    const EMPTY: Self = Self {
        index: (TimeInt::STATIC, RowId::ZERO),
        promise: None,
    };
```